pkg/apis/core/v1/types.go

```diff
@@ -76,6 +77,29 @@ type (
 		Items           []Function `json:"items"`
 	}
 
+	//FunctionStatus ...
+	FunctionStatus struct {
+		// FnStatus is the function status.
+		FnStatus BuildStatus `json:"functionstatus,omitempty"`
+
+		// EnvStatus is the environment status
+		EnvStatus BuildStatus `json:"envStatus,omitempty"`
```

Comment:
I am still not convinced we need envStatus and PkgStatus as two separate fields - tomorrow if we add more objects which are related to/used by a function - should we add more `objStatus` fields here may be not I think. I still think structure similar to https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties is better - and many design choices in controller actually end up because of these two status fields being here. And as a natural effect of removing Env & Pkg Status - the log fields should also go from here.

Reply:
I have two part response for this:
1) I am open to adopting more generic `objStatus` structure as you have suggested. I have devised current structure based on the one we are using for `Package Status` i.e. 
```
	PackageStatus struct {
		BuildStatus BuildStatus `json:"buildstatus,omitempty"`
		BuildLog string `json:"buildlog,omitempty"` // output of the build (errors etc)
		LastUpdateTimestamp metav1.Time `json:"lastUpdateTimestamp,omitempty"`
	}
```
I thought this one could be more readable for fission users as they are used to `PackageStatus` structure

Conclusion: I will devise a generic scalable `objStatus` structure for FunctionStatus and implement it in the next iteration.